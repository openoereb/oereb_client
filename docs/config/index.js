var path = require('canonical-path');

var Package = require('dgeni').Package;

module.exports = new Package('oereb_client_doc', [
  require('dgeni-packages/ngdoc'),
  require('dgeni-packages/nunjucks')
])

  .processor(require('./processors/index-page'))
  .processor(require('./processors/version'))
  .processor(require('./processors/api-data'))
  .processor(require('./processors/guide-data'))

  .config(function (log, readFilesProcessor, writeFilesProcessor) {

    // Set the log level to 'info', switch to 'debug' when troubleshooting
    log.level = 'info';

    // Specify the base path used when resolving relative paths to source and output files
    readFilesProcessor.basePath = path.resolve(__dirname, '../..');

    // Specify our source files that we want to extract
    readFilesProcessor.sourceFiles = [
      { // All of our application files
        include: 'oereb_client/static/js/**/*.js',
        basePath: 'oereb_client/static/js'
      },
      { // Our static Markdown documents
        include: 'docs/content/**/*.md',
        basePath: 'docs/content',
        fileReader: 'ngdocFileReader'
      }
    ];

    // Use the writeFilesProcessor to specify the output folder for the extracted files
    writeFilesProcessor.outputFolder = 'docs/build';

  })

  .config(function (parseTagsProcessor, getInjectables) {
    parseTagsProcessor.tagDefinitions =
      parseTagsProcessor.tagDefinitions.concat(getInjectables(require('./tag-defs')));
  })

  .config(function (templateFinder) {

    // Specify where the templates are located
    templateFinder.templateFolders.unshift(path.resolve(__dirname, 'templates'));

  })

  .config(function (computePathsProcessor, computeIdsProcessor) {

    computePathsProcessor.pathTemplates.push({
      docTypes: ['module'],
      pathTemplate: '${area}/${name}',
      outputPathTemplate: 'partials/${area}/${name}.html'
    });

    computePathsProcessor.pathTemplates.push({
      docTypes: ['componentGroup'],
      pathTemplate: '${area}/${moduleName}/${groupType}',
      outputPathTemplate: 'partials/${area}/${moduleName}/${groupType}.html'
    });

    // create new compute for 'content' type doc
    // indexPage is something new we will be defining later
    computeIdsProcessor.idTemplates.push({
      docTypes: ['content', 'indexPage'],
      getId: function (doc) {
        return doc.fileInfo.baseName;
      },
      getAliases: function (doc) {
        return [doc.id];
      }
    });

    // Build custom paths and set the outputPaths for "content" pages
    computePathsProcessor.pathTemplates.push({
      docTypes: ['content'],
      getPath: function (doc) {
        var docPath = path.dirname(doc.fileInfo.relativePath);
        if (doc.fileInfo.baseName !== 'index') {
          docPath = path.join(docPath, doc.fileInfo.baseName);
        }
        return docPath;
      },
      outputPathTemplate: 'partials/${path}.html'
    });

  })

  .config(function (extractAccessTransform) {
    extractAccessTransform.allowedDocTypes.add('js');
  });
