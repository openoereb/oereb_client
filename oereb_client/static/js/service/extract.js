goog.provide('oereb.ExtractService');

goog.require('oereb');

/**
 * Angular service to request a specific extract.
 * @param {angular.$http} $http Angular service for HTTP requests.
 * @param {angular.$q} $q Angular service for deferrable objects.
 * @param {angular.Module.constant} oerebApplicationUrl Angular service for HTTP requests.
 * @constructor
 * @ngInject
 * @ngdoc service
 * @ngname ExtractService
 */
oereb.ExtractService = function($http, $q, oerebApplicationUrl) {

  this.$http_ = $http;
  this.$q_ = $q;
  this.applicationUrl_ = oerebApplicationUrl;

  this.extract_ = undefined;
  this.embeddable_ = undefined;

};

/**
 * Requests the extract for the specified EGRID.
 * @param {string} egrid The EGRID to request the extract for.
 * @returns {angular.$q.Promise} Promise for the EGRID request.
 */
oereb.ExtractService.prototype.queryExtractById = function(egrid) {
  this.extract_ = undefined;
  this.embeddable_ = undefined;
  var def = this.$q_.defer();
  this.$http_.get(this.applicationUrl_ + 'extract/reduced/json/geometry/' + egrid).then(
    function(response) {
      if (angular.isObject(response.data['GetExtractByIdResponse'])) {
        this.extract_ = response.data['GetExtractByIdResponse']['extract'];
        if (angular.isObject(response.data['GetExtractByIdResponse']['embeddable'])) {
          this.embeddable_ = response.data['GetExtractByIdResponse']['embeddable'];
        }
        def.resolve(response.data['GetExtractByIdResponse']);
      }
      else {
        def.reject('Invalid response format.');
      }
    }.bind(this),
    function(response) {
      var error = 'Requesting extract for ' + egrid + ' failed.';
      if (angular.isString(response.data)) {
        error += ' ' + response.data;
      }
      def.reject(error);
    }
  );
  return def.promise;
};

/**
 * Returns the extract if available.
 * @returns {Object|undefined} The extract object or undefined.
 */
oereb.ExtractService.prototype.getExtract = function() {
  return this.extract_;
};

/**
 * Returns the real estate data if available.
 * @returns {Object|undefined} The real estate data or undefined.
 */
oereb.ExtractService.prototype.getRealEstate = function() {
  if (angular.isDefined(this.getExtract())) {
    return this.getExtract()['RealEstate'];
  }
  return undefined;
};

/**
 * Returns list of concerned themes.
 * @returns {Array|undefined} The list of concerned themes.
 */
oereb.ExtractService.prototype.getConcernedThemes = function() {
  if (angular.isDefined(this.getExtract())) {
    return this.getExtract()['ConcernedTheme'];
  }
  return undefined;
};

/**
 * Returns list of not concerned themes.
 * @returns {Array|undefined} The list of not concerned themes.
 */
oereb.ExtractService.prototype.getNotConcernedThemes = function() {
  if (angular.isDefined(this.getExtract())) {
    return this.getExtract()['NotConcernedTheme'];
  }
  return undefined;
};

/**
 * Returns list of themes without data.
 * @returns {Array|undefined} The list of themes without data.
 */
oereb.ExtractService.prototype.getThemesWithoutData = function() {
  if (angular.isDefined(this.getExtract())) {
    return this.getExtract()['ThemeWithoutData'];
  }
  return undefined;
};

/**
 * Returns the public law restrictions of the specified topic.
 * @param {string} themeCode The code of the topic to return the restrictions for.
 * @returns {Array|undefined} The public law restrictions of the specified topic.
 */
oereb.ExtractService.prototype.getRestrictions = function(themeCode) {
  if (angular.isDefined(this.getRealEstate())) {
    var restrictions = [];
    angular.forEach(this.getRealEstate()['RestrictionOnLandownership'], function(restriction) {
      if (restriction['Theme']['Code'] === themeCode) {
        restrictions.push(restriction);
      }
    });
    return restrictions;
  }
  return undefined;
};

/**
 * Returns the legend entries of the specified topic.
 * @param {string} themeCode The code of the topic to return the restrictions for.
 * @returns {Object|undefined} The legend entries of the specified topic.
 */
oereb.ExtractService.prototype.getLegend = function(themeCode) {
  var restrictions = this.getRestrictions(themeCode);
  if (angular.isDefined(restrictions)) {
    var legendEntries = [];
    var legendGraphics = [];
    for (var i = 0; i < restrictions.length; i++) {
      var existing = false;
      for (var j = 0; j < legendEntries.length; j++) {
        if (legendEntries[j]['TypeCode'] === restrictions[i]['TypeCode']) {
          existing = true;
          legendEntries[j]['Area'] += restrictions[i]['Area'];
          legendEntries[j]['PartInPercent'] += restrictions[i]['PartInPercent'];
          break;

        }

      }
      if (!existing) {
        legendEntries.push({
          'TypeCode': restrictions[i]['TypeCode'],
          'Information': restrictions[i]['Information'],
          'Area': restrictions[i]['Area'],
          'PartInPercent': restrictions[i]['PartInPercent'],
          'SymbolRef': restrictions[i]['SymbolRef']
        });
      }
      this.addIfNotContains_(restrictions[i]['Map']['LegendAtWeb'], legendGraphics);
    }
    return {
      'entries': legendEntries,
      'graphics': legendGraphics
    };
  }
  return undefined;
};

/**
 * Adds the item to the specified target array if it's not already contained. `angular.equals` is used for the
 * comparison.
 * @param {Object|Array|string|number} item The element to be added.
 * @param {Array} target The target array to add the element to.
 * @private
 */
oereb.ExtractService.prototype.addIfNotContains_ = function(item, target) {
  if (angular.isArray(target)) {
    for (var i = 0; i < target.length; i++) {
      if (angular.equals(target[i], item)) {
        return;
      }
    }
    target.push(item);
  }
};

/**
 * Adds the document to the specified target array if it's not already contained. `angular.equals` is used for
 * the comparison.
 * @param {Object} document The document to be added.
 * @param {Array} target The target array to add the element to.
 * @private
 */
oereb.ExtractService.prototype.addDocumentIfNotContained_ = function(document, target) {
  for (var i = 0; i < target.length; i++) {
    // Check if document already exists (only use 'Title', 'OfficialNumber' and 'TextAtWeb' for comparison)
    if (
      angular.equals(target[i]['Title'], document['Title']) &&
      angular.equals(target[i]['OfficialNumber'], document['OfficialNumber']) &&
      angular.equals(target[i]['TextAtWeb'], document['TextAtWeb'])
    ) {
      // Add missing article numbers if document already exists
      for (var j = 0; j < document['ArticleNumber'].length; j++) {
        this.addIfNotContains_(document['ArticleNumber'][j], target[i]['ArticleNumber']);
      }
      // Add missing articles if document already exists
      for (var k = 0; k < document['Article'].length; k++) {
        this.addIfNotContains_(document['Article'][k], target[i]['Article']);
      }
      return;
    }
  }
  // Add document if it is missing
  target.push(document);
};

/**
 * Add each listed document to the target array if it's not already contained.
 * @param {Array} documents The documents to be added.
 * @param {Array} target The target array.
 * @private
 */
oereb.ExtractService.prototype.addDocumentsIfNotContained_ = function(documents, target) {
  // Iterate documents to be added
  for (var i = 0; i < documents.length; i++) {
    // Create document object with necessary properties and check if it needs to be added
    this.addDocumentIfNotContained_({
      'Title': documents[i]['Title'],
      'OfficialNumber': documents[i]['OfficialNumber'],
      'ArticleNumber': documents[i]['ArticleNumber'] || [],
      'Article': documents[i]['Article'] || [],
      'TextAtWeb': documents[i]['TextAtWeb']
    }, target);
    // Do the same for possible references
    var references = documents[i]['Reference'];
    if (angular.isArray(references)) {
      this.addDocumentsIfNotContained_(references, target);
    }
  }
};

/**
 * Returns the legal documents of the specified topic.
 * @param {string} themeCode The code of the topic to return the restrictions for.
 * @returns {Object|undefined} The legal documents of the specified topic.
 */
oereb.ExtractService.prototype.getDocuments = function(themeCode) {
  // Get the restrictions for the specified topic
  var restrictions = this.getRestrictions(themeCode);
  if (angular.isArray(restrictions)) {
    var legalProvisions = [];
    var documents = [];
    // Iterate the resulting restrictions
    for (var i = 0; i < restrictions.length; i++) {
      // Iterate the legal provisions for each restriction
      var legalProvision = restrictions[i]['LegalProvisions'];
      for (var j = 0; j < legalProvision.length; j++) {
        // Create legal provision object with necessary properties and check if it needs to be added
        this.addDocumentIfNotContained_({
          'Title': legalProvision[j]['Title'],
          'OfficialNumber': legalProvision[j]['OfficialNumber'],
          'ArticleNumber': legalProvision[j]['ArticleNumber'] || [],
          'Article': legalProvision[j]['Article'] || [],
          'TextAtWeb': legalProvision[j]['TextAtWeb']
        }, legalProvisions);
        // Add possible references to the documents
        var references = legalProvision[j]['Reference'];
        if (angular.isArray(references)) {
          this.addDocumentsIfNotContained_(references, documents);
        }
      }
    }
    // Add possible references for the real estate to the documents
    this.addDocumentsIfNotContained_(this.getRealEstate()['Reference'], documents);
    return {
      'LegalProvisions': legalProvisions,
      'Documents': documents
    };
  }
  return undefined;
};

/**
 * Returns the responsible offices of the specified topic.
 * @param {string} themeCode The code of the topic to return the responsible offices for.
 * @returns {Array|undefined} The responsible offices of the specified topic.
 */
oereb.ExtractService.prototype.getResponsibleOffices = function(themeCode) {
  var restrictions = this.getRestrictions(themeCode);
  if (angular.isArray(restrictions)) {
    var offices = [];
    for (var i = 0; i < restrictions.length; i++) {
      this.addIfNotContains_(restrictions[i]['ResponsibleOffice'], offices);
    }
    return offices;
  }
  return undefined;
};

/**
 * Creates a view service definition using the restriction's theme code and view service URL.
 * @param {string} themeCode The restriction's theme code.
 * @param {string} viewServiceUrl The restriction's view service URL.
 * @returns {Object} The created view service definition.
 * @private
 */
oereb.ExtractService.prototype.getViewServiceFromUrl_ = function(themeCode, viewServiceUrl) {
  var parts = viewServiceUrl.split('?');
  var url = parts[0];
  var definition = {
    'topic': themeCode,
    'url': url,
    'params': {}
  };
  var params = parts[1].split('&');
  for (var i = 0; i < params.length; i++) {
    var param = params[i].split('=');
    if (param[0].toUpperCase() === 'LAYERS') {
      definition['params']['LAYERS'] = param[1];
    }
    else if (param[0].toUpperCase() === 'STYLES') {
      definition['params']['STYLES'] = param[1];
    }
    else if (param[0].toUpperCase() === 'VERSION') {
      definition['params']['VERSION'] = param[1];
    }
  }
  return definition;
};

/**
 * Returns a unique list of view services available in the extract.
 * @returns {Array} A unique list of view service objects.
 */
oereb.ExtractService.prototype.getViewServices = function() {
  var viewServices = [];
  var realEstate = this.getRealEstate();
  if (angular.isDefined(realEstate)) {
    var restrictions = realEstate['RestrictionOnLandownership'];
    if (angular.isArray(restrictions)) {
      for (var i = 0; i < restrictions.length; i++) {
        var url = restrictions[i]['Map']['ReferenceWMS'];
        if (angular.isDefined(url)) {
          this.addIfNotContains_(
            this.getViewServiceFromUrl_(restrictions[i]['Theme']['Code'], url),
            viewServices
          );
        }
      }
    }
  }
  return viewServices;
};

/**
 * Returns the exclusions of liability for the current extract.
 * @returns {Array} The exclusions of liability.
 */
oereb.ExtractService.prototype.getExclusionsOfLiability = function() {
  if (angular.isDefined(this.getExtract()) && angular.isArray(this.getExtract()['ExclusionOfLiability'])) {
    return this.getExtract()['ExclusionOfLiability'];
  }
  return [];
};

/**
 * Returns the glossary for the current extract.
 * @returns {Array} The exclusions of liability.
 */
oereb.ExtractService.prototype.getGlossary = function() {
  if (angular.isDefined(this.getExtract()) && angular.isArray(this.getExtract()['Glossary'])) {
    return this.getExtract()['Glossary'];
  }
  return [];
};

/**
 * Returns the embeddable if available.
 * @returns {Object|undefined} The extract object or undefined.
 */
oereb.ExtractService.prototype.getEmbeddable = function() {
  return this.embeddable_;
};

oereb.module.service('ExtractService', oereb.ExtractService);
