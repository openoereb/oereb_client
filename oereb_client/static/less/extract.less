/**
 * Extract styling
 */

@import "variables";

div#extract {
  width: @menu-with;
  max-width: 100%;
  position: fixed;
  top: 0;
  left: 0;
  height: 0;
  padding-top: @header-height;
  background-color: #fff;
  transition: height @animation-duration;
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: stretch;
  div.extract-wrapper {
    z-index: 1;
    background-color: @menu-background-color;
    overflow-y: scroll;
    flex: 1;
    padding-bottom: @default-padding;
    margin-top: @default-padding;
  }
  span.extract-close {
    z-index: 2;
    display: block;
    float: right;
    color: @brand-primary;
    opacity: 1;
  }
  span.extract-close:hover {
    color: #000000;
    opacity: 1;
  }
  .panel-default > .panel-heading .badge {
    background-color: @brand-primary;
  }
  .collapse-extract {
    padding-top: @default-padding;
    .btn {
      border-bottom-width: 0;
    }
  }
}

div#extract.loading {
  height: @loading-indicator-height;
}

div#extract.active {
  height: 100%;
}

div#extract.information {
  box-shadow: @menu-shadow;
}