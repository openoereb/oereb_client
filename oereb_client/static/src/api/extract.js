export function queryExtractById(applicationUrl, egrid) {
    const url = new URL(applicationUrl + 'extract/reduced/json/geometry/' + egrid);
    url.searchParams.append('_dc', new Date().getTime());
    return fetch(url).then(response => response.json());
}

export function sanitizeTopicCode(code) {
    return code.replace('.', '').replace('-', '').replace('_', '');
}

export function groupRestrictionsByTopic(retstrictions) {
    const restrictionsByTopic = {};
    retstrictions.forEach((restriction) => {
        const code = sanitizeTopicCode(restriction.Theme.Code);
        if (!restrictionsByTopic.hasOwnProperty(code)) {
            restrictionsByTopic[code] = [];
        }
        restrictionsByTopic[code].push(restriction);
    });
    return restrictionsByTopic;
}
