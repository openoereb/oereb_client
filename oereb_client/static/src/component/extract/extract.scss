@import "../../oereb_client.scss";

div.oereb-client-extract-wrapper {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    background-color: $menu-background-color;
    box-shadow: $default-shadow;
    overflow: hidden;
    transition: height $animation-duration linear;

    div.oereb-client-extract {
        width: 100%;
        height: 100%;
        padding-top: $header-height;
        overflow: hidden;
    }
    
    div.oereb-client-extract.data {
        .category:not(.show) {
            border: 0px;
        }
    
        & > div.accordion {
            overflow-y: scroll;
    
            button.accordion-button:focus,
            button.accordion-button:not(.collapsed) {
                box-shadow: none;
            }
        }

        @include media-breakpoint-up(xl) {
            button.collapse-extract {
                display: none !important;
            }
        }
    }
}

div.oereb-client-extract-wrapper.hidden {
    height: $header-height;
}

div.oereb-client-extract-wrapper.loading {
    height: $header-height + 100px;
}

div.oereb-client-extract-wrapper.shown {
    height: 100%;
}

div.oereb-client-extract-wrapper.collapsed {
    height: $header-height + 50px;
}

@include media-breakpoint-up(xl) {
    div.oereb-client-extract-wrapper {
        width: $menu-with;
    }
}