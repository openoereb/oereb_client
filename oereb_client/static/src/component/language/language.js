import PropTypes from 'prop-types';
import React from "react";
import {withNamespaces} from 'react-i18next';
import {useDispatch, useSelector} from "react-redux";

import {setLanguage} from "../../reducer/language";

const OerebLanguage = function({t}) {
  const dispatch = useDispatch();
  const language = useSelector((state) => state.language);
  const currentLanguage = language.current;
  const availableLanguages = language.available;

  const selectLanguage = function(lang) {
    dispatch(setLanguage(lang));
  };

  const languages = availableLanguages.map((language, key) =>
    <li key={key}>
      <button className="dropdown-item" onClick={selectLanguage.bind(this, language)}>
        {language.toUpperCase()}
      </button>
    </li>
  );

  return (
    <div>
      <button className="btn btn-outline-secondary dropdown-toggle"
        type="button"
        data-bs-toggle="dropdown"
        aria-expanded="false"
        title={t('menu.language.title')}>
        {currentLanguage.toUpperCase()}
      </button>
      <ul className="dropdown-menu">
        {languages}
      </ul>
    </div>
  );
};

OerebLanguage.propTypes = {
  t: PropTypes.func.isRequired
};

export default withNamespaces()(OerebLanguage);
