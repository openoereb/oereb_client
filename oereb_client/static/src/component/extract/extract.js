import './extract.scss';

import React from 'react';
import { useSelector } from 'react-redux';
import OerebExtractLoading from './extract_loading';
import OerebExtractError from './extract_error';
import OerebExtractData from './extract_data';

function OerebExtract(props) {
    const extract = useSelector((state) => state.extract);

    function getContent(cfg) {
        if (cfg.loading) {
            return (
                <OerebExtractLoading />
            );
        }
        else if (cfg.error) {
            return (
                <OerebExtractError />
            );
        }
        else if (cfg.visible) {
            return (
                <OerebExtractData data={extract.data.GetExtractByIdResponse.extract} />
            );
        }
        else {
            return null;
        }
    }

    const content = getContent(extract);

    return (
        <div class="oereb-client-extract">
            {content}
        </div>
    );
}

export default OerebExtract;
