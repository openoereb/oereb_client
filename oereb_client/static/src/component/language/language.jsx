import React from "react";
import {useTranslation} from "react-i18next";
import {useDispatch, useSelector} from "react-redux";

import {hideExtract, loadExtract, showError, showExtract} from "../../reducer/extract";
import {updateHistory} from "../../reducer/history";
import {setLanguage} from "../../reducer/language";
import {queryExtractById} from "../../request/extract";
import {TooManyRequestsError} from "../../util/error";
import {showWarning} from "../../reducer/message";

/**
 * This component provides the language selector.
 */
const OerebLanguage = function() {
  const {t} = useTranslation();
  const dispatch = useDispatch();
  const config = useSelector((state) => state.config).config;
  const serviceUrl = config.service_url;
  const language = useSelector((state) => state.language);
  const currentLanguage = language.current;
  const availableLanguages = language.available;
  const extract = useSelector((state) => state.extract);

  const selectLanguage = function(lang) {
    if (lang !== currentLanguage) {
      dispatch(setLanguage(lang));
      if (extract.visible) {
        dispatch(loadExtract({
          egrid: extract.egrid,
          identdn: extract.identdn,
          number: extract.number,
          zoom: false
        }));
        queryExtractById(
          serviceUrl,
          extract.egrid,
          extract.identdn,
          extract.number,
          config.extract_json_timeout,
          lang
        )
          .then((data) => {
            dispatch(showExtract(data));
            dispatch(updateHistory(data));
          })
          .catch((error) => {
            if (error instanceof TooManyRequestsError) {
              dispatch(hideExtract());
              dispatch(showWarning(t('extract.error.too_many_requests'), true));
            }
            else {
              dispatch(showError());
            }
          });
      }
    }
  };

  const languages = availableLanguages.map((language, key) =>
    <li key={key}>
      <button className="dropdown-item" onClick={selectLanguage.bind(this, language)}>
        {language.toUpperCase()}
      </button>
    </li>
  );

  return (
    <div className="oereb-client-language-selector">
      <button className="btn btn-outline-secondary dropdown-toggle"
        type="button"
        data-bs-toggle="dropdown"
        aria-expanded="false"
        title={t("menu.language.title")}>
        {currentLanguage.toUpperCase()}
      </button>
      <ul className="dropdown-menu">
        {languages}
      </ul>
    </div>
  );
};

export default OerebLanguage;
