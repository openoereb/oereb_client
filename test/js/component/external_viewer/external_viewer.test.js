import {mount} from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import {act} from "react-dom/test-utils";
import {Provider} from "react-redux";

import OerebExternalViewer
  from "../../../../oereb_client/static/src/component/external_viewer/external_viewer";
import {update} from "../../../../oereb_client/static/src/reducer/config";
import {hideExtract, showExtract} from "../../../../oereb_client/static/src/reducer/extract";
import {initLanguages} from "../../../../oereb_client/static/src/reducer/language";
import MainStore from "../../../../oereb_client/static/src/store/main";
import extract from "../../../../samples/extract.json";

describe('external viewer component', () => {

  let component;

  beforeEach(() => {
    act(() => {
      MainStore.dispatch(initLanguages({
        default: 'de',
        available: ['de']
      }));
    });
    act(() => {
      MainStore.dispatch(showExtract(extract));
    });
    act(() => {
      MainStore.dispatch(update({
        external_viewer: {
          url: 'http://example.com',
          tooltip: [
            {
              Language: 'de',
              Text: 'WebGIS'
            }
          ],
          params: [
            'egrid={egrid}'
          ]
        }
      }));
    });
    component = mount(
      <Provider store={MainStore}>
        <OerebExternalViewer />
      </Provider>
    );
    window.open = jest.fn();
  });

  afterEach(() => {
    window.open.mockClear();
  });

  it('should render button', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should not render button', () => {
    act(() => {
      MainStore.dispatch(update({}));
    });
    const wrapper = mount(
      <Provider store={MainStore}>
        <OerebExternalViewer />
      </Provider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call window.open', () => {
    component.find('button').simulate('click');
    expect(window.open.mock.calls).toHaveLength(1);
    expect(window.open.mock.calls[0][0]).toEqual('http://example.com?egrid=CH1234');
    expect(window.open.mock.calls[0][1]).toEqual('_blank');
    window.open.mockReset();
  });

  it('should not call window.open', () => {
    act(() => {
      MainStore.dispatch(hideExtract());
    });
    component.find('button').simulate('click');
    expect(window.open.mock.calls).toHaveLength(0);
    window.open.mockReset();
  });

});