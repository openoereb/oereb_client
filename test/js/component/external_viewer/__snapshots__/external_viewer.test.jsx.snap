// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`external viewer component > should not render button 1`] = `<DocumentFragment />`;

exports[`external viewer component > should render button 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outline-secondary oereb-client-external-viewer"
    title="WebGIS"
    type="button"
  >
    <i
      class="bi bi-map"
    />
  </button>
</DocumentFragment>
`;
